p2p/tools/local_network.py

```diff
@@ -0,0 +1,378 @@
+
+    def writelines(self, data: Iterable[bytes]) -> None:
+        for line in data:
+            self.write(line)
+            self.write(b'\n')
+
+    def write_eof(self) -> None:
+        self._reader.feed_eof()
+
+    def can_write_eof(self) -> bool:
+        return True
+
+    def is_closing(self) -> bool:
+        return False
+
+    def close(self) -> None:
+        self.write_eof()
+
+
+class AddressedTransport(MemoryTransport):
+    """
+    Direct connection between a StreamWriter and StreamReader.
+    """
+    _queue: asyncio.Queue
+
+    def __init__(self, address: Address, reader: asyncio.StreamReader) -> None:
+        super().__init__(reader)
+        self._address = address
+        self._queue = asyncio.Queue()
+
+    @property
+    def queue(self) -> asyncio.Queue:
+        return self._queue
+
+    def get_extra_info(self, name: str, default=None):
+        if name == 'peername':
+            return (self._address.host, self._address.port)
+        else:
+            return super().get_extra_info(name, default)
+
+    def write(self, data: bytes) -> None:
+        super().write(data)
+        self._queue.put_nowait(len(data))
+
+
+ReaderWriterPair = Tuple[asyncio.StreamReader, asyncio.StreamWriter]
+
+
+def addressed_pipe(address: Address) -> ReaderWriterPair:
+    reader = asyncio.StreamReader()
+
+    transport = AddressedTransport(address, reader)
+    protocol = asyncio.StreamReaderProtocol(reader)
+
+    writer = asyncio.StreamWriter(
+        transport=transport,
+        protocol=protocol,
+        reader=reader,
+        loop=asyncio.get_event_loop(),
+    )
+    return reader, writer
+
+
+def direct_pipe() -> ReaderWriterPair:
+    reader = asyncio.StreamReader()
+
+    transport = MemoryTransport(reader)
+    protocol = asyncio.StreamReaderProtocol(reader)
+
+    writer = asyncio.StreamWriter(
+        transport=transport,
+        protocol=protocol,
+        reader=reader,
+        loop=asyncio.get_event_loop(),
+    )
+    return reader, writer
+
+
+async def _connect_streams(
+        reader: asyncio.StreamReader,
+        writer: asyncio.StreamWriter,
+        queue: asyncio.Queue,
+        token: CancelToken) -> None:
+    try:
+        while not token.triggered:
+            if reader.at_eof():
+                break
+
+            try:
+                size = queue.get_nowait()
+            except asyncio.QueueEmpty:
+                await asyncio.sleep(0)
+                continue
+            data = await wait_with_token(reader.readexactly(size), token=token)
+            writer.write(data)
+            queue.task_done()
+            await wait_with_token(writer.drain(), token=token)
+    except OperationCancelled:
+        pass
+    finally:
+        writer.write_eof()
+
+    if reader.at_eof():
+        reader.feed_eof()
+
+
+class Router(BaseService):
+    hosts: Dict[str, 'Host']
+    networks = Dict[str, 'Network']
+    connections = Dict[CancelToken, asyncio.Task]
```

Comment:
I *think* `CancelToken` could be replaced with `asyncio.Event` in a pretty straightforward way.

Reply:
Hrm, the thing I don't think it replaces is the `wait_with_token` functionality which would have to be reproduced to keep the various `awaits` in the connection loop from hanging when things get cancelled.